File: Chronos/source/Chronos/Marshaling/Generic/UInt32Marshaler.cs
using System.IO;

namespace Chronos.Marshaling
{
    public class UInt32Marshaler : GenericMarshaler<uint>
    {
        protected override void MarshalInternal(uint value, Stream stream)
        {
            Marshal(value, stream);
        }

        protected override uint DemarshalInternal(Stream stream)
        {
            return Demarshal(stream);
        }

        public static unsafe void Marshal(uint value, Stream stream)
        {
            byte[] buffer = new byte[sizeof (uint)];
            fixed (byte* pointer = buffer)
            {
                *((uint*) pointer) = value;
            }
            stream.Write(buffer, 0, buffer.Length);
        }

        public static unsafe uint Demarshal(Stream stream)
        {
            byte[] buffer = new byte[sizeof (uint)];
            stream.Read(buffer, 0, buffer.Length);
            uint value;
            fixed (byte* pointer = buffer)
            {
                value = *((uint*) pointer);
            }
            return value;
        }
    }
}

File: chronos/src/Chronos.ProfilerAgent/UInt32Marshaler.cpp
#include "StdAfx.h"
#include "UInt32Marshaler.h"


void CUInt32Marshaler::Marshal(__uint value, CBaseStream* stream)
{
	stream->Write(&value, TypeSize::_INT32);
}

__uint CUInt32Marshaler::Demarshal(CBaseStream* stream)
{
	__uint value = 0;
	stream->Read(&value, TypeSize::_INT32);
	return value;
}
